opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/FileController.java

```diff
@@ -0,0 +1,169 @@
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+import org.opengrok.indexer.search.QueryBuilder;
+import org.opengrok.web.api.v1.filter.CorsEnable;
+import org.opengrok.web.api.v1.filter.PathAuthorized;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opengrok.indexer.index.IndexDatabase.getDocument;
+
+@Path(FileController.PATH)
+public class FileController {
+
+    public static final String PATH = "/file";
+
+    private static final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    private ArrayList<String> lines = new ArrayList<>();
+
+    static class LineDTO {
+        @JsonProperty
+        private String line;
+        @JsonProperty
+        private int number;
+
+        // for testing
+        LineDTO() {
+        }
+
+        LineDTO(String line, int num) {
+            this.line = line;
+            this.number = num;
+        }
+
+        public String getLine() {
+            return this.line;
+        }
+
+        public int getNumber() {
+            return this.number;
+        }
+    }
+
+    private static File getFile(String path, HttpServletResponse response) throws IOException {
+        if (path == null) {
+            if (response != null) {
+                response.sendError(HttpServletResponse.SC_NOT_ACCEPTABLE, "Missing path parameter");
+            }
+            return null;
+        }
+
+        File file = new File(env.getSourceRootFile(), path);
+        if (!file.isFile()) {
+            if (response != null) {
+                response.sendError(HttpServletResponse.SC_NOT_FOUND, "File not found");
+            }
+            return null;
+        }
+
+        return file;
+    }
+
+    @GET
+    @CorsEnable
+    @PathAuthorized
+    @Path("/content")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Object getContent(@Context HttpServletRequest request,
+                             @Context HttpServletResponse response,
+                             @QueryParam("path") final String path) throws IOException, ParseException {
+
+        File file = getFile(path, response);
+        if (file == null) {
+            // error already set in the response
+            return null;
+        }
+
+        Document doc;
+        if ((doc = getDocument(file)) == null) {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND, "Cannot get document for file");
+            return null;
+        }
+
+        String fileType = doc.get(QueryBuilder.T);
+        if (!fileType.equals(AbstractAnalyzer.Genre.PLAIN.typeName())) {
```

Comment:
Recommend transposing to avoid possible NPE:

`!AbstractAnalyzer.Genre.PLAIN.typeName().equals(fileType)`